### ISSUE TITLE
rubocop_todo specific behavior

### ISSUE THREAD

**Issue by deivid-rodriguez** [bbatsov/rubocop#2424]
#### Request

Right now `rubocop_todo.yml` behaves just like `.rubocop.yml`, but it has some helpful comments and it's supposed to be removed. I always expected their behavior to be a little different.

I'd like `.rubocop_todo.yml` to have an extra `Limit` parameter for each Cop, say

```yaml
Metrics/MethodLength:
  Max: 80
  Limit: 13
```

and I'd like RuboCop to fail if I exceed that number of offenses.

I would then have several options:

- Fix the issue(s) I've introduced. (Do it now)
- Rebuild my todo config to have `Limit: 14` (or whatever), so RuboCop wouldn't shout again unless I have more than the new number of offenses. (Do it later)
- Move the `Max: 80` configuration to the regular `.rubocop.yml` (Don't do it)

I would also like the `rubocop_todo.yml` to be updated when the `Limit` parameter is reduced.

#### Rationale

The configurations in `.rubocop_todo.yml` file are supposed to be removed, so I think we could help the user with that. In the previous case, the user could keep writing huge methods but Rubocop won't give any feedback, so removing the offenses will get harder and harder. If we raised a flag when limits are exceeded the user would be more aware about the kind of code (s)he is writing and whether (s)he wants to fix it.

Maybe some users won't be interested, so RuboCop could behave like this under a `--strict` flag or something.

I have no idea how difficult this would be to implement, but I thought I would throw this out and see what you guys think.

**Comment by agrimm** [comment]
Not an implementation of what you're requesting, but there's the `--exclude-limit` that excludes specific files if there's n or less files with that offense. You can then check if the offense has happened in new files since the last `--auto-gen-config`.

**Comment by jonas054** [comment]
@deivid-rodriguez I think it's a good idea to try to prevent "offense creep", that offenses are added without anyone noticing just because `.rubocop_todo.yml` temporarily allows them.

There are three distinct cases here. The cops that get an `Enabled: false` parameter in the todo list, the ones that get a `Max:` parameter, and the ones that get some other setting because the code consistently follows another style, e.g. `EnforcedStyle: outdent`.

The last case is usually no problem. The deviation from the default configuration was intended, and the generated parameter will be moved to `.rubocop.yml` as permanent configuration.

The `Enabled: false` parameters are a bit of a problem. They turn off the cops entirely, so we have to make major changes if we want to detect when a configured offense count limit is exceeded. On the other hand, we have the `Exclude` parameter generation that @agrimm mentioned, so the problem is partly solved already.

The cops that get a `Max` parameter generated are the ones where the proposed change makes the most sense, and also where it can be implemented with a reasonable amount of work. My thinking is that we don't need a `--strict` flag. This is something that most people will appreciate.

For the cops that have a `Max` parameter and a `Limit` setting, we keep track of number of offenses relative to the default configuration, and start reporting offenses when the `Limit` is surpassed. Adding something to the message to clarify that the offense is reported for that reason.

**Comment by deivid-rodriguez** [comment]
Thanks guys for your explanations. I didn't fully understand the "exclude trick", but my request was focused only on cops with the `Max` parameter, since this is the only kind of cops I have in my `.rubocop_todo.yml` files. So I'm very happy this is the easy case to solve :smiley:

**Comment by jonas054** [comment]
It's the _easiest_. I wouldn't go so far as to say it's easy. :smile:

Then there's the question how well the feature will play with other features, `--only`, `--except`, caching, and so on. It's difficult to judge, could be messy. But anyway, it's worth taking a closer look and try to implement something.

**Comment by deivid-rodriguez** [comment]
Understood! :wink: 

I'll definitely give this a try if I have some time in the future!

**Comment by alexdowad** [comment]
Maintainers, is this `Limit` parameter something we want to implement? To me, it seems to add extra complexity for marginal gain. I would suggest that we can do without it.

**Comment by jonas054** [comment]
Yes, let's not keep it open forever if no-one's working on it.

**Comment by deivid-rodriguez** [comment]
If this is an accepted feature request, why closing it? :disappointed:

By the way, thanks @alexdowad for all the work you've been doing recently.

**Comment by jonas054** [comment]
It's due to lack of activity. Someone must want to do an implementation. If I got an indication that something is in the works, I'd re-open straight away.

**Comment by deivid-rodriguez** [comment]
I don't agree with it but if that's your policy you should probably close any feature request older than 2 months having no activity as well.

**Comment by bbatsov** [comment]
You'll notice that we closed most older feature requests already.

**Comment by alexdowad** [comment]
> You'll notice that we closed most older feature requests already.

Except for some which will definitely be implemented, probably soon.

**Comment by deivid-rodriguez** [comment]
I don't want to get in the middle of your work with this small thing. I'd like just confirmation that the feature request is still accepted since I said I would dedicate some time in the future to it.

I feel @alexdowad and you guys have different visions of what closing an issue means. Whereas he seems to be suggesting to reject the feature, you seem to be just closing for inactivity.

**Comment by alexdowad** [comment]
> I don't want to get in the middle of your work with this small thing. I'd like just confirmation that the feature request is still accepted since I said I would dedicate some time in the future to it.
> 
> I feel @alexdowad and you guys have different visions of what closing an issue means. Whereas he seems to be suggesting to reject the feature, you seem to be just closing for inactivity.

@deivid-rodriguez, those are two sides of the same coin. I see the implementation of this feature as problematic, and don't want to work on it for that reason. There doesn't seem to be anybody else who is working on it either.

Closing the issue doesn't mean that the feature will never be added. It just means that the contributors don't want to be reminded about it now. If you come up with an implementation and open a PR, others will certainly be happy to review.

Regarding whether the feature is "accepted", I can't speak for the repo owners, but I doubt that they can give you a definite "yes" or "no" without seeing an implementation.

**Comment by deivid-rodriguez** [comment]
Ok, don't want to keep discussing this. I'd rather let you work on `RuboCop` :wink:
